### ISSUE TITLE
Login verification failed: TypeError: Cannot read property 'close' of undefined

### ISSUE THREAD

**Issue by IonicOwl** [Alex-Rose/fb-messenger-cli#151]
Just updated to the latest version (old version just stopped logging me in, spitting out a load of errors in the process), and it won't even take my credentials now.

```
user@host:~$ fb-messenger-cli
Login verification failed: Warning: Settings not found, saving default values
Facebook credentials:
Email: HIDDEN
Password: ************
Attempting login...
Login verification failed: No saved profile, please login
Facebook credentials:
Email: HIDDEN
Password: ************
Attempting login...
Error saving .settings file
Error while logging in Error: spawn EACCES
Login verification failed: TypeError: Cannot read property 'close' of undefined
Facebook credentials:
Email: 
```

I can't even get out of this, I have to close the terminal and open a new one.  I'm happy to do a full install from nothing, but I can't figure out how to uninstall to even do that.

**Comment by geezuh** [comment]
Same issue. the npm install did pull phantomjs too, which I thought was removed. It downloaded chromium as well.

```
npm install -g fb-messenger-cli                                                                                           
/home/gzuh/.npm-global/bin/fb-messenger-cli -> /home/gzuh/.npm-global/lib/node_modules/fb-messenger-cli/cli.js                           
+ fb-messenger-cli@2.0.2                                                                                                                 
updated 1 package in 6.388s                                                                                                              
gzuh@nosuch:~% fb-messenger-cli                                                                                                          
Login verification failed: Warning: Settings not found, saving default values                                                            
Facebook credentials:                                                                                                                    
Email: [redacted]                                                                                                                  
Password: *********                                                                                                                      
Attempting login...                                                                                                                      
Login verification failed: No saved profile, please login           
```

**Comment by jakobkogler** [comment]
Same issue here: 

    Password: ***********************
    Attempting login...
    Error while logging in Error: spawn EACCES
    Login verification failed: TypeError: Cannot read property 'close' of undefined

**Comment by DrnglVrgs** [comment]
Same here

**Comment by pokyah** [comment]
Here also with node v6.13.0 and npm 3.10.10 under ubuntu 16.04

**Comment by KenN7** [comment]
same here, archlinux, npm 5.6, node 9.5

**Comment by SamBergeron** [comment]
Hi, can you guys try starting with the "noSandbox" option set to true?
So just edit the settings file `.settings` and set   `"noSandbox": true`, you could also put `headlessLogin` to false if you're running in an environment where you can visually open chrome, you'll see what's going on.

Let me know if this changes anything.

**Comment by KenN7** [comment]
I managed to launch the app, the problem seems to come from the permissions to create the .settings file, the directory /usr/lib/node_modules/fb-messenger-cli/ is owned by root and thus the app is not allowed to write there unless you launch it as root. Still launching as root is not a good solution.

**Comment by IonicOwl** [comment]
Which settings file is this?

`stuart@StuBuntu:~$ locate fb-messenger-cli | grep settings
/usr/lib/node_modules/fb-messenger-cli/node_modules/puppeteer/.local-chromium/linux-515411/chrome-linux/xdg-settings
/usr/lib/node_modules/fb-messenger-cli/node_modules/puppeteer/.local-chromium/linux-515411/chrome-linux/resources/inspector/settings
/usr/lib/node_modules/fb-messenger-cli/node_modules/puppeteer/.local-chromium/linux-515411/chrome-linux/resources/inspector/settings/settings_module.js
/usr/lib/node_modules/fb-messenger-cli/scripts/settings.js
`

`Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.3 LTS
Release:	16.04
Codename:	xenial
`

**Comment by KenN7** [comment]
Has the log says, the .settings file is not saved,
> Error saving .settings file

The file gets created if you launch the app in sudo and is located in
> /usr/lib/node_modules/fb-messenger-cli/.settings

**Comment by MagnusRufum** [comment]
I can confirm that "noSandbox" was set to false by default and changing to true allowed me to log in. However it still required to launch as root.

**Comment by SamBergeron** [comment]
Ahh not having access to create the ,settings file actually makes a lot of sense.
I'll work on a solution, because indeed, forcing root for the first login is not ideal.

... that is the workaround meanwhile. 
Can you guys confirm that once the first login is done with root and the .settings file is created, you can log in normally without root?

**Comment by MagnusRufum** [comment]
After the first login with root, I can login without root.

**Comment by googleson78** [comment]
I could not login even with root. Chromium(me) complained about starting without sandbox as root. It also did not create a default settings file, even as root.

**Comment by slashome** [comment]
Tried to login as root :

```bash
Login verification failed: Warning: Settings not found, saving default values
Facebook credentials:
Email: <login>
Password: ********
Attempting login...
Login verification failed: No saved profile, please login
Facebook credentials:
Email: <login>
Password: ********
Attempting login...
Settings have been saved
Error while logging in Error: Failed to launch chrome!
[0327/142218.225711:ERROR:zygote_host_impl_linux.cc(90)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.


TROUBLESHOOTING: https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md

Login verification failed: TypeError: Cannot read property 'close' of undefined
Facebook credentials:
Email:
```

Can I do anything to help ?

**Archlinux** kernel 4.15.12-1-ARCH
**Node** v8.9.1
**NPM** 5.5.1

**Comment by SamBergeron** [comment]
@googleson78, @slashome, this says you can't run as root without --no-sandbox. There's currently an option in the .settings file where you can start without sandbox. Could you try with that turned on? I might have to make it default.

**Comment by slashome** [comment]
@SamBergeron I tried without success :/

```bash
** ⚡ root@<machine>  /home/<user>  fb-messenger-cli --no-sandbox**
Login verification failed: Warning: Settings not found, saving default values
Facebook credentials:
Email: <login>
Password: ********
Attempting login...
Login verification failed: No saved profile, please login
Facebook credentials:
Email: <login>
Password: ********
Attempting login...
Settings have been saved
Error while logging in Error: Failed to launch chrome!
[0327/164702.132865:ERROR:zygote_host_impl_linux.cc(90)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.


TROUBLESHOOTING: https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md

Login verification failed: TypeError: Cannot read property 'close' of undefined
Facebook credentials:
Email: 
```

**Comment by SamBergeron** [comment]
@slashome Sorry that's not what I meant, you have to start puppeteer with --no-sandbox and not fb-messenger-cli. So we've made a helper option for that:
Go to fb-messenger-cli install directory,
Modify the `.settings` file for that the option `noSandbox` is set to `true`
Then start the app again, it'll pass that option to Puppeteer when it needs to.

**Comment by slashome** [comment]
Oh sorry I get it now !
After a root login I can log without root. Please let us now when this manipulation will be not necessary anymore so I can reinstall it the clean way ;)
Thank for your work

**Comment by boardfish** [comment]
Running it with sudo doesn't generate .settings. Using nvm with node v9.10.1 on Windows Subsystem for Linux (Ubuntu).

**Comment by yukenz** [comment]
1.First try run _fb-messenger-cli_ for normal user after you install it with _sudo npm install -g fb-messenger-cli --unsafe-perm_ ,and you will get it
![fbcli1](https://user-images.githubusercontent.com/34228964/40877474-adf1759e-66ab-11e8-97f6-5950b5e82efe.jpeg)
2.Try running with sudo,so like it _sudo fb-messenger-cli_ and you will see error like it and you will get **.settings** file in **/usr/lib/node_modules/fb-messenger-cli/.settings**
![fbcli2](https://user-images.githubusercontent.com/34228964/40877493-e39e7890-66ab-11e8-8eea-a7550cd9e4bc.jpeg)
3.edit **.settings** file in path above and set headless to false and no sandbox true like it
![fbcli3](https://user-images.githubusercontent.com/34228964/40877502-1e36b90e-66ac-11e8-9819-15f9d8e359c7.jpeg)
4.after this (optional) if you have bad connection ,you will not able to login this time because login,js only give chrome time login 40 second so you should edit **/usr/lib/node_modules/fb-messenger-cli/scripts/login.js** and find word "could not login this time" and change variable on above this text like it (recommend 100-200).
![fbcli4](https://user-images.githubusercontent.com/34228964/40877518-8ba68c44-66ac-11e8-9519-2bfdd79e7a55.jpeg)
5.Try again login with sudo (_sudo fb-messenger-cli_) and you will se chromium has open,dont worry it will close automatic after success login.
![fbcli5](https://user-images.githubusercontent.com/34228964/40877520-9fb79cdc-66ac-11e8-9a5d-2b150621bf2d.jpeg)
6.you able to login
![fbcli6](https://user-images.githubusercontent.com/34228964/40877530-c1b41c02-66ac-11e8-9bf8-09754a3211fe.jpeg)
7.now you can start without sudo
![fbcli7](https://user-images.githubusercontent.com/34228964/40877535-ccaa4884-66ac-11e8-9a99-e2761a617fef.jpeg)

After time login has expired you will login again with sudo,and if you update your fb-messenger-cli may you need to reconfigure it again.
I hope it work for you.

**Comment by Mshanmuganathan** [comment]
I too have the same problem. Login verification failed: TypeError: Cannot read property 'close' of undefined

**Comment by Mshanmuganathan** [comment]
@Alex-Rose Help us with this issue! It's been open for 5 months!

**Comment by ex0dus-0x** [comment]
Here's what I have done to alleviate this issue.

(I'm currently on `x86_64 Arch Linux`)

After installing with `npm`, I launch with `sudo`:

```
$ sudo fb-messenger-cli
```

This is when I presume the `.settings` file is checked for existence and created. Attempting to log in failed for me, as `fb-messenger-cli` needed to launch `chromium` in order to input my information. I killed the client and edited `/usr/lib/node_modules/fb-messenger-cli/.settings` as so:

```
{
  "disableColors": false,
  "groupColors": true,
  "conversationsToLoad": 15,
  "preventMessageFlicker": false,
  "desktopNotifications": false,
  "showTimestamps": false,
  "useCustomNicknames": true,
  "timestampLocale": "en-US",
  "timestampOptions": {},
  "twoFactorAuth": false,
  "headlessLogin": true,
  "noSandbox": true,
  "logonTimeout": 86400000
}
```

Only changed made was the `noSandbox` value set to true.

**Comment by IonicOwl** [comment]
Ok I think I got this working, at least when running with sudo...

I was able to tab-complete my way to `/usr/lib/node_modules/fb-messenger-cli/.settings`, even though I couldn't find it before, but when opening the file, I got nothing - it was empty.  I did attempt to populate the file with the contents above, but I was unable to save it, so I opened the file with sudo (like `sudo vi /usr/lib/node_modules/fb-messenger-cli/.settings`), the managed to succesfully save it.  

EDIT:  It looks like it works fine without sudo now.  I'm guessing I still needed to run it once as root, so I'm not sure that part is fixed yet.
